core/src/main/java/org/jruby/truffle/translator/BodyTranslator.java

```diff
@@ -346,9 +351,139 @@ public RubyNode visitBreakNode(org.jruby.ast.BreakNode node) {
     }
 
+    private RubyNode translateRubiniusPrimitive(SourceSection sourceSection, CallNode node) {
+        /*
+         * Translates something that looks like
+         *
+         *   Rubinius.primitive :foo
+         *
+         * into
+         *
+         *   CallRubiniusPrimitiveNode(FooNode(arg1, arg2, ..., argN))
+         *
+         * Where the arguments are the same arguments as the method. It looks like this is only exercised with simple
+         * arguments so we're not worrying too much about what happens when they're more complicated (rest,
+         * keywords etc).
+         */
+
+        if (node.getArgsNode().childNodes().size() != 1 || !(node.getArgsNode().childNodes().get(0) instanceof org.jruby.ast.SymbolNode)) {
+            throw new UnsupportedOperationException("Rubinius.primitive must have a single literal symbol argument");
+        }
+
+        final String primitiveName = ((org.jruby.ast.SymbolNode) node.getArgsNode().childNodes().get(0)).getName();
+
+        final RubiniusPrimitiveConstructor primitive = context.getRubiniusPrimitiveManager().getPrimitive(primitiveName);
+
+        final List<RubyNode> arguments = new ArrayList<>();
+
+        int argumentsCount = primitive.getFactory().getExecutionSignature().size();
+
+        if (primitive.getAnnotation().needsSelf()) {
+            arguments.add(new SelfNode(context, sourceSection));
+            argumentsCount--;
+        }
+
+        for (int n = 0; n < argumentsCount; n++) {
+            arguments.add(new ReadPreArgumentNode(context, sourceSection, n, MissingArgumentBehaviour.UNDEFINED));
+        }
+
+        return new CallRubiniusPrimitiveNode(context, sourceSection,
+                primitive.getFactory().createNode(context, sourceSection, arguments.toArray(new RubyNode[arguments.size()])),
+                environment.getReturnID());
+    }
+
+    private RubyNode translateRubiniusInvokePrimitive(SourceSection sourceSection, CallNode node) {
+        /*
+         * Translates something that looks like
+         *
+         *   Rubinius.invoke_primitive :foo, arg1, arg2, argN
+         *
+         * into
+         *
+         *   CallRubiniusPrimitiveNode(FooNode(arg1, arg2, ..., argN))
+         */
+
+        if (node.getArgsNode().childNodes().size() < 1 || !(node.getArgsNode().childNodes().get(0) instanceof org.jruby.ast.SymbolNode)) {
+            throw new UnsupportedOperationException("Rubinius.invoke_primitive must have at least an initial literal symbol argument");
+        }
+
+        final String primitiveName = ((org.jruby.ast.SymbolNode) node.getArgsNode().childNodes().get(0)).getName();
+
+        final RubiniusPrimitiveConstructor primitive = context.getRubiniusPrimitiveManager().getPrimitive(primitiveName);
+
+        final List<RubyNode> arguments = new ArrayList<>();
+
+        for (org.jruby.ast.Node arg : node.getArgsNode().childNodes().subList(1, node.getArgsNode().childNodes().size())) {
```

Comment:
In general, this looks like a very Ruby-style way of expressing this. Requesting an iterator and skipping with next over the first is more efficient in Java.